### PR TITLE
Fix Store caching

### PR DESCRIPTION
**Pull request by mikee47**
This PR fixes the issues raised in #29.

The performance tests have also been revised as setValue / getValue shouldn't include opening store.

**Add second cache for writes**

The database keeps a weak reference to all store instances used for updates. The main purpose for this is to determine whether a store is locked or not.

In addition, there is now a second write cache (a strong reference) so that subsequent updates don't have to reload from storage.

**Add `StoreRef` and `StoreUpdateRef` classes**

`StoreRef` inherits from `shared_ptr<Store>` and tracks both read and update use of a Store.
The destructor calls into the database so it can effectively decide when caches are no longer required and the memory can be freed.
Note that this has to take asynchronous updates into consideration.

A `StoreUpdateRef` inherits from `StoreRef` and handles reference-counting directly in its constructor/destructor/assignment operator.
This avoids calling `commit` in the Store destructor - there be dragons.

**openStore**

The `Database::openStore` method now just opens a store for reading. It returns a `StoreRef` instance.

To get an updatable store requires a second call to `Database::lockStore`. This returns a `StoreUpdateRef` instance. On failure, it returns a dummy instance which avoids dereferenced null pointers later on. The `bool` operator for this instance returns false.

The `openStoreForUpdate` is provided which does both operations together. One use-case for this is `database.openStoreForUpdate(0)->clear();` which is a one-liner to reset a store to defaults and commit it.

To summarise behaviour:

- Call OPEN to obtain read-only store
- Call LOCK to get updatable store. Operation results in either a COPY of an existing store, a LOAD from storage or FAIL.
- Can COPY where store is locked (in-use) but clean (not dirty), otherwise we need to LOAD a fresh copy from storage.

### CODE REVIEW

**Comment by mikee47** [comment]
This PR fixes the issues raised in #29.

The performance tests have also been revised as setValue / getValue shouldn't include opening store.

**Add second cache for writes**

The database keeps a weak reference to all store instances used for updates. The main purpose for this is to determine whether a store is locked or not.

In addition, there is now a second write cache (a strong reference) so that subsequent updates don't have to reload from storage.


**Add `StoreRef` and `StoreUpdateRef` classes**

`StoreRef` inherits from `shared_ptr<Store>` and tracks both read and update use of a Store.
The destructor calls into the database so it can effectively decide when caches are no longer required and the memory can be freed.
Note that this has to take asynchronous updates into consideration.

A `StoreUpdateRef` inherits from `StoreRef` and handles reference-counting directly in its constructor/destructor/assignment operator.
This avoids calling `commit` in the Store destructor - there be dragons.


**openStore**

The `Database::openStore` method now just opens a store for reading. It returns a `StoreRef` instance.

To get an updatable store requires a second call to `Database::lockStore`. This returns a `StoreUpdateRef` instance. On failure, it returns a dummy instance which avoids dereferenced null pointers later on. The `bool` operator for this instance returns false.

The `openStoreForUpdate` is provided which does both operations together. One use-case for this is `database.openStoreForUpdate(0)->clear();` which is a one-liner to reset a store to defaults and commit it.

To summarise behaviour:

- Call OPEN to obtain read-only store
- Call LOCK to get updatable store. Operation results in either a COPY of an existing store, a LOAD from storage or FAIL.
- Can COPY where store is locked (in-use) but clean (not dirty), otherwise we need to LOAD a fresh copy from storage.
